### ISSUE TITLE
Very Slow Access to blockDim.y and gridDim.y

### ISSUE THREAD

**Issue by fthaler** [ROCm-Developer-Tools/HIP#1283]
We observe a strange performance degradation when using `hipBlockDim.y` and possibly `hipGridDim.y` in even very simple HIP kernels. For demonstration, a basic copy-bandwidth estimation code is provided. It basically measures the performance of linear memory copy (no fancy indexing), once using the index `i = blockDim.x * blockIdx.x + threadIdx.x`, once with `i = blockDim.y * blockIdx.y + threadIdx.y` and once with `i = blockDim.z * blockIdx.z + threadIdx.z`. Block sizes are adapted accordingly. See the full code:

```c++
#include <algorithm>
#include <chrono>
#include <functional>
#include <iostream>
#include <random>
#include <vector>

#ifdef __HIPCC__
#include <hip/hip_runtime.h>
#define GPU(expr) hip##expr
#elif defined(__CUDACC__)
#define GPU(expr) cuda##expr
#endif

inline void gpu_check(GPU(Error_t) error, const std::string &file, int line) {
  if (error != GPU(Success)) {
    std::cerr << "GPU error: " << GPU(GetErrorName(error)) << " at " << file
              << ":" << line << std::endl;
    std::abort();
  }
}

#define GPU_CALL(expr) gpu_check(GPU(expr), __FILE__, __LINE__)

using copy_t = double;

__global__ void copy_x(copy_t *dst, const copy_t *src, int n) {
  for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < n;
       i += blockDim.x * gridDim.x)
    dst[i] = src[i];
}

__global__ void copy_y(copy_t *dst, const copy_t *src, int n) {
  for (int i = blockIdx.y * blockDim.y + threadIdx.y; i < n;
       i += blockDim.y * gridDim.y)
    dst[i] = src[i];
}

__global__ void copy_z(copy_t *dst, const copy_t *src, int n) {
  for (int i = blockIdx.z * blockDim.z + threadIdx.z; i < n;
       i += blockDim.z * gridDim.z)
    dst[i] = src[i];
}

template <class Kernel, class... Args>
double benchmark(Kernel &&kernel, dim3 const &block_count,
                 dim3 const &block_size, Args &&... args) {
  using clock = std::chrono::high_resolution_clock;
  std::vector<double> timings;

  for (int i = 0; i < 100; ++i) {
    auto start = clock::now();
    kernel<<<block_count, block_size>>>(std::forward<Args>(args)...);
    GPU_CALL(GetLastError());
    GPU_CALL(DeviceSynchronize());
    auto stop = clock::now();
    timings.push_back(std::chrono::duration<double>(stop - start).count());
  }

  std::nth_element(timings.begin(), timings.begin() + timings.size() / 2,
                   timings.end());
  return timings[timings.size() / 2];
}

int main(int argc, char *argv[]) {
  if (argc != 2) {
    std::cout << "Usage: " << argv[0] << " ELEMENTS" << std::endl;
    return 1;
  }

  const int n = std::atoi(argv[1]);
  std::cout << "number of elements: " << n << std::endl;

  copy_t *src, *dst;
  GPU_CALL(Malloc(&src, sizeof(copy_t) * n));
  GPU_CALL(Malloc(&dst, sizeof(copy_t) * n));

  std::vector<copy_t> hsrc(n), hdst(n);

  std::minstd_rand engine;
  std::uniform_real_distribution<copy_t> distribution;
  auto random = std::bind(distribution, std::ref(engine));

  int device;
  GPU_CALL(GetDevice(&device));
#ifdef __HIPCC__
  hipDeviceProp_t device_props;
#else
  cudaDeviceProp device_props;
#endif
  GPU_CALL(GetDeviceProperties(&device_props, device));
  const int block_size = device_props.maxThreadsPerBlock;

  dim3 block_size_x(std::min(block_size, device_props.maxThreadsDim[0]), 1, 1);
  dim3 block_count_x(std::min(int((n + block_size_x.x - 1) / block_size_x.x),
                              device_props.maxGridSize[0]),
                     1, 1);

  dim3 block_size_y(1, std::min(block_size, device_props.maxThreadsDim[1]), 1);
  dim3 block_count_y(1,
                     std::min(int((n + block_size_y.y - 1) / block_size_y.y),
                              device_props.maxGridSize[1]),
                     1);

  dim3 block_size_z(1, 1, std::min(block_size, device_props.maxThreadsDim[2]));
  dim3 block_count_z(1, 1,
                     std::min(int((n + block_size_z.z - 1) / block_size_z.z),
                              device_props.maxGridSize[2]));

  const double gigabytes = sizeof(copy_t) * n * 2 / 1.0e9;

  std::generate(hsrc.begin(), hsrc.end(), random);
  GPU_CALL(
      Memcpy(src, hsrc.data(), sizeof(copy_t) * n, GPU(MemcpyHostToDevice)));
  auto seconds = benchmark(copy_x, block_count_x, block_size_x, dst, src, n);
  std::cout << "copy_x estimated bandwidth: " << (gigabytes / seconds)
            << " GB/s" << std::endl;
  GPU_CALL(
      Memcpy(hdst.data(), dst, sizeof(copy_t) * n, GPU(MemcpyDeviceToHost)));
  if (hdst != hsrc) {
    std::cerr << "copy_x did not work!" << std::endl;
    std::abort();
  }

  std::generate(hsrc.begin(), hsrc.end(), random);
  GPU_CALL(
      Memcpy(src, hsrc.data(), sizeof(copy_t) * n, GPU(MemcpyHostToDevice)));
  seconds = benchmark(copy_y, block_count_y, block_size_y, dst, src, n);
  std::cout << "copy_y estimated bandwidth: " << (gigabytes / seconds)
            << " GB/s" << std::endl;
  GPU_CALL(
      Memcpy(hdst.data(), dst, sizeof(copy_t) * n, GPU(MemcpyDeviceToHost)));
  if (hdst != hsrc) {
    std::cerr << "copy_x did not work!" << std::endl;
    std::abort();
  }

  std::generate(hsrc.begin(), hsrc.end(), random);
  GPU_CALL(
      Memcpy(src, hsrc.data(), sizeof(copy_t) * n, GPU(MemcpyHostToDevice)));
  seconds = benchmark(copy_z, block_count_z, block_size_z, dst, src, n);
  std::cout << "copy_y estimated bandwidth: " << (gigabytes / seconds)
            << " GB/s" << std::endl;
  GPU_CALL(
      Memcpy(hdst.data(), dst, sizeof(copy_t) * n, GPU(MemcpyDeviceToHost)));
  if (hdst != hsrc) {
    std::cerr << "copy_z did not work!" << std::endl;
    std::abort();
  }

  GPU_CALL(Free(src));
  GPU_CALL(Free(dst));

  return 0;
}
```

Note that the macros are just there to directly allow for CUDA compilation with NVCC. If we now compile this code and run it on a Vega Mi50/GFX906, we get the following performance numbers (with 10000000 elements, bandwidth is slightly higher with more elements):
```bash
$ ./a.out 10000000
number of elements: 10000000
copy_x estimated bandwidth: 769.638 GB/s
copy_y estimated bandwidth: 117.511 GB/s
copy_y estimated bandwidth: 768.049 GB/s
```

The `copy_y` performance is obviously terrible, while the others are essentially the same (what we would hope and expect). With another GFX900 card, we get a similar effect, although less strong:
```bash
$ ./a.out 10000000
number of elements: 10000000
copy_x estimated bandwidth: 349.467 GB/s
copy_y estimated bandwidth: 195.286 GB/s
copy_y estimated bandwidth: 352.656 GB/s
```

On an NVIDIA V100, we don’t see this effect:
```
$ ./a.out 10000000
number of elements: 10000000
copy_x estimated bandwidth: 755.18 GB/s
copy_y estimated bandwidth: 755.144 GB/s
copy_y estimated bandwidth: 754.859 GB/s
```

We have analyzed the optimized LLVM bitcode, it is identical for all the three kernels (apart from the direction indices of course). However, there are obvious differences in the generated assembly, though we are not sure what’s going wrong and what’s in the end decreasing the performance so much. EDIT: the LLVM bitcode is not _exactly_ identical, the alignment of the 16bit values is different and only 2 bytes for the y direction, which seems to trigger the generation of different load instructions.

If you replace the `blockDim.y` and `gridDim.y` by anything else, e.g. compile time constants, `copy_y` is as fast as the others. So we assume it’s not a hardware problem.

For completeness, the full disassembled code of the fast `copy_x` and the slow `copy_y` (`copy_z` has the same code as `copy_x` apart from register naming).

```
0000000000002000 copy_x(double*, double const*, int):
	s_load_dword s0, s[4:5], 0x4                               // 000000002000: C0020002 00000004
	s_load_dword s3, s[4:5], 0xc                               // 000000002008: C00200C2 0000000C
	s_load_dword s2, s[6:7], 0x10                              // 000000002010: C0020083 00000010
	s_mov_b32 s10, s9                                          // 000000002018: BE8A0009
	s_mov_b32 s32, s10                                         // 00000000201C: BEA0000A
	s_waitcnt lgkmcnt(0)                                       // 000000002020: BF8CC07F
	s_and_b32 s4, s0, 0xffff                                   // 000000002024: 8604FF00 0000FFFF
	s_mul_i32 s0, s8, s4                                       // 00000000202C: 92000408
	s_sub_i32 s0, s3, s0                                       // 000000002030: 81800003
	s_min_u32 s5, s0, s4                                       // 000000002034: 83850400
	s_mul_i32 s0, s5, s8                                       // 000000002038: 92000805
	v_add_u32_e32 v0, s0, v0                                   // 00000000203C: 68000000
	v_cmp_gt_i32_e32 vcc, s2, v0                               // 000000002040: 7D880002
	s_and_saveexec_b64 s[0:1], vcc                             // 000000002044: BE80206A
	s_cbranch_execz BB0_3                                      // 000000002048: BF88003C

000000000000204c BB0_1:
	v_cvt_f32_u32_e32 v1, s4                                   // 00000000204C: 7E020C04
	v_rcp_iflag_f32_e32 v1, v1                                 // 000000002050: 7E024701
	v_mul_f32_e32 v1, 0x4f800000, v1                           // 000000002054: 0A0202FF 4F800000
	v_cvt_u32_f32_e32 v1, v1                                   // 00000000205C: 7E020F01
	v_mul_lo_u32 v2, v1, s4                                    // 000000002060: D2850002 00000901
	v_mul_hi_u32 v3, v1, s4                                    // 000000002068: D2860003 00000901
	v_sub_u32_e32 v4, 0, v2                                    // 000000002070: 6A080480
	v_cmp_eq_u32_e32 vcc, 0, v3                                // 000000002074: 7D940680
	v_cndmask_b32_e32 v2, v2, v4, vcc                          // 000000002078: 00040902
	v_mul_hi_u32 v2, v2, v1                                    // 00000000207C: D2860002 00020302
	v_add_u32_e32 v3, v1, v2                                   // 000000002084: 68060501
	v_sub_u32_e32 v1, v1, v2                                   // 000000002088: 6A020501
	v_cndmask_b32_e32 v1, v1, v3, vcc                          // 00000000208C: 00020701
	v_mul_hi_u32 v1, v1, s3                                    // 000000002090: D2860001 00000701
	v_mul_lo_u32 v2, v1, s4                                    // 000000002098: D2850002 00000901
	v_add_u32_e32 v3, 1, v1                                    // 0000000020A0: 68060281
	v_add_u32_e32 v4, -1, v1                                   // 0000000020A4: 680802C1
	v_sub_u32_e32 v5, s3, v2                                   // 0000000020A8: 6A0A0403
	v_cmp_ge_u32_e32 vcc, s3, v2                               // 0000000020AC: 7D9C0403
	v_cmp_le_u32_e64 s[0:1], s4, v5                            // 0000000020B0: D0CB0000 00020A04
	s_and_b64 s[0:1], s[0:1], vcc                              // 0000000020B8: 86806A00
	v_cndmask_b32_e64 v1, v1, v3, s[0:1]                       // 0000000020BC: D1000001 00020701
	v_cndmask_b32_e32 v1, v4, v1, vcc                          // 0000000020C4: 00020304
	v_mul_lo_u32 v2, v1, s4                                    // 0000000020C8: D2850002 00000901
	s_mov_b64 s[0:1], 0                                        // 0000000020D0: BE800180
	v_cmp_gt_u32_e32 vcc, s3, v2                               // 0000000020D4: 7D980403
	v_addc_co_u32_e32 v1, vcc, 0, v1, vcc                      // 0000000020D8: 38020280
	v_mul_lo_u32 v2, v1, s5                                    // 0000000020DC: D2850002 00000B01
	s_load_dwordx4 s[4:7], s[6:7], 0x0                         // 0000000020E4: C00A0103 00000000

00000000000020ec BB0_2:
	v_ashrrev_i32_e32 v1, 31, v0                               // 0000000020EC: 2202009F
	v_lshlrev_b64 v[3:4], 3, v[0:1]                            // 0000000020F0: D28F0003 00020083
	s_waitcnt lgkmcnt(0)                                       // 0000000020F8: BF8CC07F
	v_mov_b32_e32 v1, s7                                       // 0000000020FC: 7E020207
	v_add_co_u32_e32 v5, vcc, s6, v3                           // 000000002100: 320A0606
	v_addc_co_u32_e32 v6, vcc, v1, v4, vcc                     // 000000002104: 380C0901
	global_load_dwordx2 v[5:6], v[5:6], off                    // 000000002108: DC548000 057F0005
	v_mov_b32_e32 v1, s5                                       // 000000002110: 7E020205
	v_add_co_u32_e32 v3, vcc, s4, v3                           // 000000002114: 32060604
	v_addc_co_u32_e32 v4, vcc, v1, v4, vcc                     // 000000002118: 38080901
	v_add_u32_e32 v0, v0, v2                                   // 00000000211C: 68000500
	v_cmp_le_i32_e32 vcc, s2, v0                               // 000000002120: 7D860002
	s_or_b64 s[0:1], vcc, s[0:1]                               // 000000002124: 8780006A
	s_waitcnt vmcnt(0)                                         // 000000002128: BF8C0F70
	global_store_dwordx2 v[3:4], v[5:6], off                   // 00000000212C: DC748000 007F0503
	s_andn2_b64 exec, exec, s[0:1]                             // 000000002134: 89FE007E
	s_cbranch_execnz BB0_2                                     // 000000002138: BF89FFEC

000000000000213c BB0_3:
	s_endpgm                                                   // 00000000213C: BF810000
```

```
0000000000002200 copy_y(double*, double const*, int):
	v_mov_b32_e32 v2, s4                                       // 000000002200: 7E040204
	v_mov_b32_e32 v3, s5                                       // 000000002204: 7E060205
	global_load_ushort v2, v[2:3], off offset:6                // 000000002208: DC488006 027F0002
	s_load_dword s3, s[4:5], 0x10                              // 000000002210: C00200C2 00000010
	s_load_dword s2, s[6:7], 0x10                              // 000000002218: C0020083 00000010
	s_mov_b32 s11, s10                                         // 000000002220: BE8B000A
	s_mov_b32 s32, s11                                         // 000000002224: BEA0000B
	s_waitcnt vmcnt(0)                                         // 000000002228: BF8C0F70
	v_mul_lo_u32 v0, s9, v2                                    // 00000000222C: D2850000 00020409
	s_waitcnt lgkmcnt(0)                                       // 000000002234: BF8CC07F
	v_sub_u32_e32 v0, s3, v0                                   // 000000002238: 6A000003
	v_min_u32_e32 v3, v0, v2                                   // 00000000223C: 1C060500
	v_mul_lo_u32 v0, v3, s9                                    // 000000002240: D2850000 00001303
	v_add_u32_e32 v0, v0, v1                                   // 000000002248: 68000300
	v_cmp_gt_i32_e32 vcc, s2, v0                               // 00000000224C: 7D880002
	s_and_saveexec_b64 s[0:1], vcc                             // 000000002250: BE80206A
	s_cbranch_execz BB1_3                                      // 000000002254: BF88003C

0000000000002258 BB1_1:
	v_cvt_f32_u32_e32 v1, v2                                   // 000000002258: 7E020D02
	s_load_dwordx4 s[4:7], s[6:7], 0x0                         // 00000000225C: C00A0103 00000000
	v_rcp_iflag_f32_e32 v1, v1                                 // 000000002264: 7E024701
	v_mul_f32_e32 v1, 0x4f800000, v1                           // 000000002268: 0A0202FF 4F800000
	v_cvt_u32_f32_e32 v1, v1                                   // 000000002270: 7E020F01
	v_mul_lo_u32 v4, v1, v2                                    // 000000002274: D2850004 00020501
	v_mul_hi_u32 v5, v1, v2                                    // 00000000227C: D2860005 00020501
	v_sub_u32_e32 v6, 0, v4                                    // 000000002284: 6A0C0880
	v_cmp_eq_u32_e32 vcc, 0, v5                                // 000000002288: 7D940A80
	v_cndmask_b32_e32 v4, v4, v6, vcc                          // 00000000228C: 00080D04
	v_mul_hi_u32 v4, v4, v1                                    // 000000002290: D2860004 00020304
	v_add_u32_e32 v5, v1, v4                                   // 000000002298: 680A0901
	v_sub_u32_e32 v1, v1, v4                                   // 00000000229C: 6A020901
	v_cndmask_b32_e32 v1, v1, v5, vcc                          // 0000000022A0: 00020B01
	v_mul_hi_u32 v1, v1, s3                                    // 0000000022A4: D2860001 00000701
	v_mul_lo_u32 v4, v1, v2                                    // 0000000022AC: D2850004 00020501
	v_add_u32_e32 v5, 1, v1                                    // 0000000022B4: 680A0281
	v_add_u32_e32 v6, -1, v1                                   // 0000000022B8: 680C02C1
	v_sub_u32_e32 v7, s3, v4                                   // 0000000022BC: 6A0E0803
	v_cmp_ge_u32_e32 vcc, s3, v4                               // 0000000022C0: 7D9C0803
	v_cmp_ge_u32_e64 s[0:1], v7, v2                            // 0000000022C4: D0CE0000 00020507
	s_and_b64 s[0:1], s[0:1], vcc                              // 0000000022CC: 86806A00
	v_cndmask_b32_e64 v1, v1, v5, s[0:1]                       // 0000000022D0: D1000001 00020B01
	v_cndmask_b32_e32 v1, v6, v1, vcc                          // 0000000022D8: 00020306
	v_mul_lo_u32 v2, v1, v2                                    // 0000000022DC: D2850002 00020501
	s_mov_b64 s[0:1], 0                                        // 0000000022E4: BE800180
	v_cmp_gt_u32_e32 vcc, s3, v2                               // 0000000022E8: 7D980403
	v_addc_co_u32_e32 v1, vcc, 0, v1, vcc                      // 0000000022EC: 38020280
	v_mul_lo_u32 v2, v1, v3                                    // 0000000022F0: D2850002 00020701

00000000000022f8 BB1_2:
	v_ashrrev_i32_e32 v1, 31, v0                               // 0000000022F8: 2202009F
	v_lshlrev_b64 v[3:4], 3, v[0:1]                            // 0000000022FC: D28F0003 00020083
	s_waitcnt lgkmcnt(0)                                       // 000000002304: BF8CC07F
	v_mov_b32_e32 v1, s7                                       // 000000002308: 7E020207
	v_add_co_u32_e32 v5, vcc, s6, v3                           // 00000000230C: 320A0606
	v_addc_co_u32_e32 v6, vcc, v1, v4, vcc                     // 000000002310: 380C0901
	global_load_dwordx2 v[5:6], v[5:6], off                    // 000000002314: DC548000 057F0005
	v_mov_b32_e32 v1, s5                                       // 00000000231C: 7E020205
	v_add_co_u32_e32 v3, vcc, s4, v3                           // 000000002320: 32060604
	v_addc_co_u32_e32 v4, vcc, v1, v4, vcc                     // 000000002324: 38080901
	v_add_u32_e32 v0, v0, v2                                   // 000000002328: 68000500
	v_cmp_le_i32_e32 vcc, s2, v0                               // 00000000232C: 7D860002
	s_or_b64 s[0:1], vcc, s[0:1]                               // 000000002330: 8780006A
	s_waitcnt vmcnt(0)                                         // 000000002334: BF8C0F70
	global_store_dwordx2 v[3:4], v[5:6], off                   // 000000002338: DC748000 007F0503
	s_andn2_b64 exec, exec, s[0:1]                             // 000000002340: 89FE007E
	s_cbranch_execnz BB1_2                                     // 000000002344: BF89FFEC

0000000000002348 BB1_3:
	s_endpgm                                                   // 000000002348: BF810000
```

Hope, you can help us fixing the performance problem!

**Comment by gargrahul** [comment]
@fthaler  Are you compiling using hip-clang?

**Comment by fthaler** [comment]
Oh totally forgot to mention, sorry! Yes, it’s hip-clang built from sources:
```
$ clang --version
clang version 10.0.0 (clang cee670adc4bf7d2337b61f7daa7821cb1b21e8c8) (llvm 05f47c7bc8cf346bae77b8cd40058a45791d2108)
Target: x86_64-unknown-linux-gnu
Thread model: posix
``

**Comment by fthaler** [comment]
It’s actually exactly the same behavior on HIP HCC (ROCm 2.6), just replace the CUDA-like kernel call in the provided example by `hipLaunchKernelGGL` to test.

**Comment by fthaler** [comment]
Any progress on this?

The source of the problem seems to be the alignment of the 16bit values of blockDim.y and gridDim.y. While *Dim.x and *Dim.z are aligned to 4, respectively 8 bytes (at least that’s what’s specified in the generated LLVM bitcode), *Dim.y is only aligned to 2 bytes, which leads to the generation of a word_load_ushort instruction instead of a s_load_dword as for the other two dimensions.

**Comment by fthaler** [comment]
This is fixed in ROCm 3.7.0 and newer versions.
